File: Tutor Demos/Week 6/Java/Team.java
import java.util.ArrayList;
import java.util.Random;

public class Team {
    private final String localName;
    private final String teamName;
    private final ArrayList<Integer> results;
    private final Random random;

    public Team(String name) {
        this.localName = name.split(" ")[0];
        this.teamName = name.split(" ")[1];
        this.results = new ArrayList<>();
        this.random = new Random();
    }

    public int play(Team opponent) {
        int result = random.nextInt(91) * (random.nextBoolean() ? 1 : -1);
        this.addResult(result);
        opponent.addResult(result*-1);
        return result;
    }

    private void addResult(int result) {
        results.add(result);
    }

    public int best() {
        if (noWins()) {
            return 0;
        }
        int max = Integer.MIN_VALUE;
        for (Integer result : results) {
            if (result > max) {
                max = result;
            }
        }
        return max;
    }

    public int worst() {
        if (noLosses()) {
            return 0;
        }
        int min = Integer.MAX_VALUE;
        for (Integer result : results) {
            if (result < min) {
                min = result;
            }
        }
        return min;
    }

    private boolean noWins() {
        for (int result : results) {
            if (result > 0) {
                return false;
            }
        }
        return true;
    }

    private boolean noLosses() {
        for (int result : results) {
            if (result < 0) {
                return false;
            }
        }
        return true;
    }

    public boolean has(String substring) {
        return localName.toLowerCase().contains(substring.toLowerCase()) || teamName.toLowerCase().contains(substring.toLowerCase());
    }

    public double average() {
        if (results.isEmpty()) {
            return 0;
        }
        int sum = 0;
        for (Integer result : results) {
            sum += result;
        }
        return (double)sum/results.size();
    }

    @Override
    public String toString() {
        return localName + " " + teamName;
    }
}

File: Tutor Demos/Week 6/Python/team.py
import random

class Team:
    def __init__(self, name):
        self.__local_name = name.split()[0]
        self.__team_name = name.split()[1]
        self.__results = []
    
    def play(self, opponent):
        result = random.randint(1, 91) * (1 if random.choice([True, False]) else -1)
        self.__results.append(result)
        opponent.__results.append(result*-1)
        return result

    def has(self, substring):
        return substring.lower() in self.__local_name.lower() or substring.lower() in self.__team_name.lower()

    def best(self):
        if self.no_wins():
            return 0
        highest = float('-inf')
        for result in self.__results:
            if result > highest:
                highest = result
        return highest

    def worst(self):
        if self.no_losses():
            return 0
        smallest = float('inf')
        for result in self.__results:
            if result < smallest:
                smallest = result
        return smallest
    
    def average(self):
        if len(self.__results) == 0:
            return 0
        total = 0
        for result in self.__results:
            total += result
        return total/len(self.__results)

    def no_wins(self):
        for result in self.__results:
            if result >= 0:
                return False
        return True

    def no_losses(self):
        for result in self.__results:
            if result < 0:
                return False
        return True

    def __str__(self):
        return f"{self.__local_name} {self.__team_name}"